### PR TITLE
Transformer + 1D/2D/3D Vision Transformer

### PR DESCRIPTION
**Pull request by avihu111**

### CODE REVIEW

File: fuse/dl/models/backbones/backbone_vit.py [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+    Projects a 1D/2D/3D image into tokens, and then runs it through a transformer
+    """
+
+    def __init__(self, token_dim: int, projection_kwargs: dict, transformer_kwargs: dict):
+        super().__init__()
+        self.projection_layer = ProjectPatchesTokenizer(token_dim=token_dim, **projection_kwargs)
+        num_tokens = self.projection_layer.num_tokens
+        self.transformer = Transformer(num_tokens=num_tokens, token_dim=token_dim, **transformer_kwargs)
+
+    def forward(self, x: torch.Tensor, pool: str = "none"):
```
Comment (mosheraboh) [inline fuse/dl/models/backbones/backbone_vit.py:65]: doc got args

**Review by mosheraboh — CHANGES_REQUESTED** [review]
Thanks Avihu! Looks very good!
Please just add documentation to any not straightforward argument.
And move the main() to be a simple unittest

**Comment by mosheraboh** [comment]
The jenkinks tests fail in all PRs. Sagi and I looking for a solution. You can ignore it for now.

File: fuse/dl/models/backbones/backbone_vit.py [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+        x = self.projection_layer(x)
+        x = self.transformer(x)
+        if pool == "cls":
+            x = x[:, 0]
+        if pool == "mean":
+            x = x.mean(dim=1)
+        return x
+
+
+def main():
```
[the remark this answers is not in the document]
Comment (avihu111) [reply fuse/dl/models/backbones/backbone_vit.py]: I changed it to be in the format of a code example.. think it's a nice code snippet.

File: fuse/dl/models/backbones/backbone_vit.py [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+        x = x.transpose(1, 2)  # x.shape == (batch_size, num_tokens, token_dim)
+        return x
+
+
+class Vit(nn.Module):
+    """
+    Projects a 1D/2D/3D image into tokens, and then runs it through a transformer
+    """
+
+    def __init__(self, token_dim: int, projection_kwargs: dict, transformer_kwargs: dict):
```
[the remark this answers is not in the document]
Comment (avihu111) [reply fuse/dl/models/backbones/backbone_vit.py:59]: think it is better now

**Review by mosheraboh — APPROVED** [review]

**Comment by egozi** [comment]
I want to add a few things:
1. default parameters for depth, heads etc.
2. standard models like - vit_base, vit_large, …
3. the interpolation module for diff image sizes.
4. What about pre trained weights? and an option flag.

**Comment by SagiPolaczek** [comment]
@egozi 
Hey Amir :) Sounds cool.

Could you please open an issue on those suggestions?
While its best to open with our templates, there is a quicker way to do so:  In the comment's menu click "Reference in a new issue".

**Pull request merged**